### ISSUE TITLE
zipkin-instrumention-request callbacks have the wrong traceId

### ISSUE THREAD

**Issue by mkante** [openzipkin/zipkin-js#179]
I setup two microservices, `server1` makes request to `server2`. The goal of the simulation is to prove that the async callbacks are executed with the wrong traceIDs (the callbacks are not bounded to CLS)

Here is my project structure 
```
/zipkin-test
  /package.json
  /service1.js
  /service2.js
```
The package.js content
```
{
  "name": "zipkin-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "async": "^2.6.0",
    "body-parser": "^1.18.2",
    "express": "^4.16.2",
    "memcached": "^2.2.2",
    "redis": "^2.8.0",
    "request": "^2.83.0",
    "request-promise": "^4.2.2",
    "zipkin": "^0.11.2",
    "zipkin-context-cls": "^0.11.0",
    "zipkin-instrumentation-express": "^0.11.2",
    "zipkin-instrumentation-memcached": "^0.11.2",
    "zipkin-instrumentation-request": "^0.11.2"
  }
}
```
The service1.js content
```
var Tracer = require('zipkin').Tracer;
var ConsoleRecorder = require('zipkin').ConsoleRecorder;
var CLSContext = require('zipkin-context-cls');
var request = require('request-promise');
var app = require('express')();
var wrapRequest = require('zipkin-instrumentation-request');
var zipkinMiddleware = require('zipkin-instrumentation-express').expressMiddleware;

var log = console.log;

// Create a tracer
var tracer = new Tracer({
  ctxImpl: new CLSContext('zipkin'),
  traceId128Bit: true,
  recorder: new ConsoleRecorder(function(){}),
  localServiceName: 'service-1'
});

log('Root TraceID: ', tracer.id.traceId);

var zipkinRequest = wrapRequest(request, { tracer: tracer });

// Install zipkin express middleware
app.use(zipkinMiddleware({ tracer: tracer }));

app.get('/service1/:id', function(req, res){
  var requestId = req.params.id;
  log('Incoming request trace ID: ' + tracer.id.traceId, ' req:', requestId);
  zipkinRequest({
    url: 'http://localhost:3001/service2',
    body: {
      reqId: requestId,
    },
    json: true,
  }).then(function(body) {
    log('Callback TraceID: ', tracer.id.traceId, ' req:', requestId);
    res.status(200).send(body);
  })
  .catch(function(err) {
    log('ERROR: Callback TraceID: ', tracer.id.traceId, ' req:', requestId);
    res.status(500).send(err)
  });
});

app.listen(3000);
```
The service2.js content
```
var request = require('request-promise');
var app = require('express')();
var bodyParser = require('body-parser')

var log = console.log;

app.use(bodyParser.json());
app.get('/service2', function(req, res){
  log(req.body);
  var reqId = req.body.reqId;

  if (reqId[0] === '1') {
    // Do not send the response, let the connect hanging
    return;
  }

  setTimeout(function() {
    res.status(200).send(['banana', 'apple']);
  }, 4000);
});

app.listen(3001);
```
Now start both services
`node service1.js`
`node service2.js`
Make a concurrent requests to service1
```
curl http://localhost:3000/service1/100 &
curl http://localhost:3000/service1/110 &
curl http://localhost:3000/service1/200 &
curl http://localhost:3000/service1/201 &
curl http://localhost:3000/service1/202 &
curl http://localhost:3000/service1/203 &
curl http://localhost:3000/service1/204 

# outputs:
Root TraceID:  4ca0ca74517e39c1a731ea2f3e203f02
Incoming request trace ID: 192de3c396e1f65862a83e8597905f6e  req: 110
Incoming request trace ID: f753e8dd73934ded5257aa0a51ac1f9d  req: 100
Incoming request trace ID: 4932ece6a19007ebf8a1123d07aba8a3  req: 200
Incoming request trace ID: af1f9df9ba00249d5d5a1257dff25c18  req: 201
Incoming request trace ID: bffab228c78f9b9201573ae4a7c96541  req: 202
Incoming request trace ID: 80cac2e2e45040bd8b03d8a51305d817  req: 203
Incoming request trace ID: 2a8bfeab02b72654adb756e8c15ade4a  req: 204
Callback TraceID:  4932ece6a19007ebf8a1123d07aba8a3  req: 200
Callback TraceID:  af1f9df9ba00249d5d5a1257dff25c18  req: 201
Callback TraceID:  af1f9df9ba00249d5d5a1257dff25c18  req: 202
Callback TraceID:  af1f9df9ba00249d5d5a1257dff25c18  req: 203
Callback TraceID:  2a8bfeab02b72654adb756e8c15ade4a  req: 204
ERROR: Callback TraceID:  f753e8dd73934ded5257aa0a51ac1f9d  req: 110
ERROR: Callback TraceID:  f753e8dd73934ded5257aa0a51ac1f9d  req: 100
...
```
As you can see the callback traceIds are wrong, `req 100` should be `192de3c396e1f65862a83e8597905f6e`, `req 202` should be `bffab228c78f9b9201573ae4a7c96541`

My bug may be similar to #176

**Comment by mkante** [comment]
Everything works fine if I use callback pattern
```
zipkinRequest({
    url: 'http://localhost:3001/service2',
    body: {
      reqId: requestId,
    },
    json: true,
  }, function(err, body, resp) {
    if (err) {
      log('ERROR: Callback TraceID: ', tracer.id.traceId, ' req:', requestId);
      res.status(500).send(err)
      return;
    }
    log('Callback TraceID: ', tracer.id.traceId, ' req:', requestId);
    res.status(200).send(body);
  });
..
```

**Comment by mkante** [comment]
To confirm my previous comment, the `request-promise` module will not work, since the promise callbacks are executed out of the original scope. Normally you want all yours callbacks running in the correct scope with the right traceId. I have been trying few hacks to make it work but they are little ugly. Maybe my expectations are too high, I will keep searching for a clean solution.

**Comment by adriancole** [comment]
thanks for the update. Not sure if code in Tracer.local can help with
regards to this or not (from approach pov once a hook is found)

**Comment by mkante** [comment]
Sorry, for the delay It's been a crazy last week :). So to respond to your comment `Tracer.local` is not exactly the solution, since it will create a new child span. Updating a matured code base will be 
The best solution I have is to force the promises to restore the `originalId` before every callback execution. 
You can extend the `Promise` class then overwrite the `then, catch` methods. Unfortunately it will not work for `then().then()...` (chaining callbacks), since `then` will returned the original `Promise` instance. So I decided to create a custom `Promise`

**Comment by mkante** [comment]
I found the solution for the issue check my [PR](https://github.com/openzipkin/zipkin-js/pull/189). I didn't want to change the existing  `zipkin-instrumentation-request` pattern so I made a new directory `zipkin-instrumentation-request-promise`. It will solve the problem for both request and request-promise.
Currently all zipkinjs instrumentation extensions  will lose the trace context if a promise is use (request-promise or redis client promisified). So I made a new `Promise` module to handle the promise callbacks trace context restoration. I was thinking maybe we could move the module into the core `zipkin` module.
This is the current solution we are using and it works well for our services.
